### PR TITLE
Start the merger of originsrv into the API

### PR DESCRIPTION
**Pull request by elliott-davis**
This only implements the channel_list endpoint as a starting off point.
We should be able to move function by function from here on out and slowly remove routes from originsrv.

Signed-off-by: Elliott Davis <elliott@excellent.io>

### CODE REVIEW

**Comment by thesentinels** [comment]
Thanks for the pull request! Here is what will happen next:
 1. Your PR will be reviewed by the maintainers
 2. If everything looks good, one of them will `approve` it, and your PR will be merged.

Thank you for contributing!

File: components/builder-api/src/server/mod.rs [hunk 1]
```diff
@@ -143,8 +152,10 @@ pub fn run(config: Config) -> Result<()> {
         cfg.listen_port()
     );
 
+    let addr = init(config.datastore.clone());
+
     server::new(move || {
-        let app_state = AppState::new(&config);
+        let app_state = AppState::new(&config, addr.clone());
```
Comment (chefsalim) [inline components/builder-api/src/server/mod.rs]: Is this creating N different datastore pools?

File: components/builder-api/src/server/models/channel.rs [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+#[table_name = "origin_channels"]
+pub struct Channel {
+    pub id: i64,
+    pub origin_id: i64,
+    pub owner_id: i64,
+    pub name: String,
+    pub created_at: Option<NaiveDateTime>,
+    pub updated_at: Option<NaiveDateTime>,
+}
+#[derive(Debug, Deserialize, Insertable)]
```
Comment (chefsalim) [inline components/builder-api/src/server/models/channel.rs]: whitesp

File: components/builder-api/src/server/resources/origins.rs [hunk 3]
```diff
@@ -1117,11 +1117,7 @@ fn download_content_as_file(content: &[u8], filename: String) -> HttpResponse {
             http::header::CONTENT_DISPOSITION,
             ContentDisposition {
                 disposition: DispositionType::Attachment,
-                parameters: vec![DispositionParam::Filename(
-                    Charset::Iso_8859_1,          // The character set for the bytes of the filename
-                    None, // The optional language tag (see `language-tag` crate)
-                    filename.as_bytes().to_vec(), // the actual bytes of the filename
-                )],
+                parameters: vec![DispositionParam::Filename(filename.clone())],
```
Comment (chefsalim) [inline components/builder-api/src/server/resources/origins.rs]: nice!

File: components/builder-originsrv/src/data_store.rs [hunk 4]
```diff
@@ -1555,33 +1555,6 @@ impl DataStore {
         occ
     }
 
-    pub fn list_origin_channels(
```
Comment (chefsalim) [inline components/builder-originsrv/src/data_store.rs]: I am wondering if we should wait to strip things out of originsrv till later... may make it easier to roll back api if needed, and only change one service at a time..

**Comment by elliott-davis** [comment]
```
[34][default:/src:0]# test/builder-api/test.sh

> hab-api-tests@0.0.1 mocha /src/test/builder-api
> mocha --require src/helpers --reporter spec src/api.js



  Origin API
    Create neurosis origin
      ✓ requires authentication
      ✓ returns the created origin
    Get origin neurosis
      ✓ returns the origin
    Create xmen origin
      ✓ returns the created origin
    Updating origins
      ✓ requires authentication
      ✓ requires that you are a member of the origin being updated
      ✓ succeeds
      ✓ reflects the changes when viewing it again
    Origin secret creation
      ✓ requires authentication
    Origin secret list
      ✓ requires authentication
      ✓ succeeds
    Origin secret deletion
      ✓ requires authentication

  Authenticate API
    Create sessions
      ✓ returns bobo
      ✓ returns mystique

  Origin Invitations API
    Invite bobo to xmen
      ✓ requires authentication
      ✓ refuses invitations from non-members
      ✓ returns the invitation
      ✓ shows bobo in the origins list of invitations
      ✓ shows xmen in bobos list of invitations
    Bobo accepts the invitation to the xmen
      ✓ requires authentication to accept the invitation
      ✓ accepts the invitation
      ✓ does not show bobo in the origins list of invitations
      ✓ does not show xmen in bobos list of invitations
      ✓ shows xmen in bobos origins

  Related Origin API functions
    Retrieving origin users
      ✓ requires authentication
      ✓ requires that you are a member of the origin
      ✓ returns the list of users
    Removing a member from an origin
      ✓ requires authentication
      ✓ requires that you are the owner of the origin
      ✓ succeeds
      ✓ no longer shows the removed user as part of the origin

  Working with packages
    Uploading packages
      ✓ does not allow unauthenticated users to upload packages
      ✓ requires that you are a member of the origin to upload a package
      ✓ allows authenticated users to upload packages (63ms)
      ✓ uploads a second package (65ms)
      ✓ uploads a third package (65ms)
      ✓ uploads a fourth package (63ms)
    Finding packages
      ✓ allows me to search for packages
      ✓ lists all packages
      ✓ lists all unique package names
      ✓ lists all packages with the specified name
      ✓ lists all versions of the package with the specified name
      ✓ returns the latest release of a package with the specified name
      ✓ lists all packages with the specified name and version
      ✓ returns the latest release of a package with the spcified name and version
      ✓ returns the specified release
    Other functions
      ✓ lists all the channels a package is in
      ✓ downloads a package
      ✓ toggles the privacy setting for a package
      ✓ requires authentication to view private packages
      ✓ does not let members of other origins view private packages
      ✓ allows members of the origin to view private packages when they are authenticated
      ✓ returns package stats

  Channels API
    Create foo channel
      ✓ requires authentication to create a channel
      ✓ returns the created channel
    Create bar channel
      ✓ succeeds
    Channel promotion
      ✓ requires authentication to promote a package
      ✓ requires origin membership to promote a package
      ✓ puts the specified package into the specified channel (79ms)
      ✓ can promote private packages
    Listing packages in a channel
      ✓ returns all packages in a channel
      ✓ returns all packages with the given name in a channel
      ✓ returns all packages with the given name and version in a channel
      ✓ returns the package with the given name, version and release in a channel (151ms)
      ✓ returns the latest package with the given name in a channel (153ms)
      ✓ returns the latest package with the given name and version in a channel (157ms)
      ✓ requires authentication to view private packages in a channel (66ms)
      ✓ does not let members of other origins view private packages in a channel (71ms)
      ✓ allows members of the origin to view private packages when they are authenticated (177ms)
    Listing channels in an origin
      ✓ returns a list of channels
    Channel demotion
      ✓ requires authentication to demote a package
      ✓ requires origin membership to demote a package
      ✓ removes the specified package from the specified channel (41ms)
      ✓ rejects attempts to demote a package from the unstable channel
      ✓ will not find a package in a channel after it has been demoted
    Delete foo channel
      ✓ rejects attempts to delete the stable channel
      ✓ rejects attempts to delete the unstable channel
      ✓ requires authentication to delete a channel
      ✓ requires origin membership to delete a channel
      ✓ succeeds

  Keys API
    Uploading public keys
      ✓ requires authentication
      ✓ requires membership in the origin you are uploading to
      ✓ uploads the key
    Downloading public keys
      ✓ can download a specific revision
      ✓ can download the latest key
    Uploading secret keys
      ✓ requires authentication
      ✓ requires membership in the origin you are uploading to
      ✓ uploads the key
    Downloading secret keys
      ✓ requires authentication
      ✓ requires membership in the origin you are uploading to
      ✓ can download the latest key
    Generating keys
      ✓ requires authentication
      ✓ requires membership in the origin you are uploading to
      ✓ generates the key
    Listing keys
      ✓ can list all public keys

  Integrations API
    Creating an origin integration
      ✓ requires authentication
      ✓ requires membership in the given origin
      ✓ succeeds
    Retrieving all integrations for an origin
      ✓ requires authentication
      ✓ requires membership in the given origin
      ✓ succeeds
    Retrieving all names for a specific integration
      ✓ requires authentication
      ✓ requires membership in the given origin
      ✓ succeeds
    Removing an integration
      ✓ requires authentication
      ✓ requires membership in the given origin
      ✓ succeeds
      ✓ creates another integration so that future tests will pass

  Profile API
    Updating the profile
      ✓ requires authentication
      ✓ allows someone to update their profile
    Retrieving a profile
      ✓ requires authentication
      ✓ allows someone to retrieve their profile

  Projects API
    Creating a project
      ✓ requires authentication
      ✓ requires membership in the origin that the project refers to
      ✓ requires a properly formatted payload
      ✓ succeeds
    Retrieving a project
      ✓ requires authentication
      ✓ requires membership in the origin that the project refers to
      ✓ succeeds
    Listing all projects in an origin
      ✓ requires authentication
      ✓ requires membership in the origin that the project refers to
      ✓ succeeds
    Editing a project
      ✓ requires authentication
      ✓ requires membership in the origin that the project refers to
      ✓ requires a properly formatted payload
      ✓ succeeds
      ✓ reflects the new changes when viewing it again
    Toggling the privacy of a project
      ✓ requires authentication
      ✓ requires membership in the origin that the project refers to
      ✓ requires that you set it to a known visibility setting
      ✓ does not allow you to set hidden manually
      ✓ succeeds
    Deleting a project
      ✓ requires authentication
      ✓ requires membership in the origin that the project refers to
      ✓ succeeds
      ✓ creates another project so that other tests dont fail

  Project integrations API
    Creating a project integration
      ✓ requires authentication
      ✓ requires membership in the origin that the project refers to
      ✓ requires a JSON body
      ✓ succeeds
    Retrieving a project integration
      ✓ requires authentication
      ✓ requires membership in the origin that the project refers to
      ✓ succeeds
    Deleting a project integration
      ✓ requires authentication
      ✓ requires membership in the origin that the project refers to
      ✓ succeeds

  Jobs API
    Scheduling jobs
      ✓ requires authentication
      ✓ requires that you belong to the origin
      ✓ does not schedule a build for Windows
      ✓ does not schedule a build for Linux kernel2
      ✓ returns the group (632ms)
    Retrieving information about a job group
      ✓ requires a group id that is a u64
      ✓ return the group
    Retrieving information about every job group in an origin
      ✓ returns all of the groups
    Listing all jobs for a project
      ✓ requires authentication
      ✓ requires membership in the origin that the project refers to
      ✓ succeeds
    Getting details of a job
      ✓ requires authentication
      ✓ requires you are a member of the origin that the job belongs to
      ✓ requires a job id that is a u64
      ✓ succeeds
    Getting logs of a job
      - requires authentication
      - requires you are logged in
      - requires a job id that is a u64
      - succeeds
      private projects
        - sets the project to private first
        - requires you are a member of the origin that the job belongs to when viewing logs for a private project
        - shows the logs for a private project
        - set the project back to public
    Promoting a job group
      ✓ requires authentication
      - requires you are a member of the origin that the job group belongs to
    Demoting a job group
      ✓ requires authentication
    Canceling a job group
      ✓ requires authentication
      ✓ requires you are a member of the origin that the job group belongs to
      ✓ cancels the group (44ms)

  External API
    Getting the external repo content
      ✓ requires authentication
      - succeeds
    Validate credentials in an external registry
      ✓ requires authentication
      - succeeds

  Miscellanenous API
    Retrieving the API status
      ✓ succeeds
    Receiving a GitHub webhook
      - succeeds
    Retrieving reverse dependencies
      ✓ returns all reverse dependencies for an origin and package name


  170 passing (3s)
  12 pending
```

File: components/builder-api/src/lib.rs [hunk 5]
```diff
@@ -15,16 +15,23 @@
...
 #[macro_use]
 extern crate bitflags;
 extern crate builder_core as bldr_core;
 extern crate bytes;
+extern crate chrono;
 extern crate constant_time_eq;
 #[macro_use]
+extern crate diesel;
+#[macro_use]
+extern crate diesel_migrations;
```
[the remark this answers is not in the document]
Comment (elliott-davis) [reply components/builder-api/src/lib.rs]: yes, but I can remove them from originsrv if you want. They are basically all doing the same thing and it should be safe

File: components/builder-api/src/server/mod.rs (hunk 1 above)
Comment (elliott-davis) [reply components/builder-api/src/server/mod.rs]: No, just a single pool inside the main actix reactor

**Comment by elliott-davis** [comment]
The integration tests will fail here for the same reason they failed with memcache - we have to build and promote a new api to get Habitat to pick up the new bindings

File: components/builder-api/src/server/db.rs [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    type Context = SyncContext<Self>;
+}
+
+pub fn init(config: DataStoreCfg) -> Addr<DbPool> {
+    let manager = ConnectionManager::<PgConnection>::new(config.to_string());
+    let pool = Pool::builder()
+        .max_size(config.pool_size)
+        .build(manager)
+        .expect("Failed to create pool.");
+    SyncArbiter::start(4, move || DbPool(pool.clone()))
```
Comment (chefsalim) [inline components/builder-api/src/server/db.rs]: config.db_workers instead of 4?

File: components/builder-api/src/config.rs [hunk 7]
```diff
@@ -80,6 +82,7 @@ impl Default for Config {
             ui: UiCfg::default(),
             upstream: UpstreamCfg::default(),
             memcached: MemcacheCfg::default(),
+            datastore: DataStoreCfg::default(),
```
Comment (chefsalim) [inline components/builder-api/src/config.rs]: would be good to also add a test of datastorecfg in the tests section below ...

File: components/builder-api/src/server/db.rs (hunk 6 above)
Comment (elliott-davis) [reply components/builder-api/src/server/db.rs]: I want to add this to builder-db but can't until we shift things around so we don't have so many reliant services on the current config structure.

**Pull request merged**